### ISSUE TITLE
x86 codegen, AVX2: missing optimization -- the compiler could swap args of commutative intrinsic `_mm256_testz_si256`

### ISSUE THREAD

**Issue by AlexGuteniev** [llvm/llvm-project#88958]
The below function has better codegen when `MASK_FIRST` is defined:
```C++
#include <immintrin.h>

bool f2(const char* p)
{
    for (const char* e = p + 256; p != e; p+=32)
    {
        const __m256i m = _mm256_loadu_si256(reinterpret_cast<const __m256i*>(p));
        if (_mm256_testz_si256(
#ifdef MASK_FIRST
                _mm256_set1_epi8(~0x40), m
#else            
                m, _mm256_set1_epi8(~0x40)
#endif
                ))
        {
            return true;
        }
    }
    
return false;
}
```
See https://godbolt.org/z/qjqdo5x69

I want the compiler to swap arguments automatically for commutative two-args operations.

**Comment by topperc** [comment]
I think the tricky bit is that it isn't commutable if the C flag is used.

**Comment by AlexGuteniev** [comment]
> I think the tricky bit is that it isn't commutable if the C flag is used.

Yes, but I think the only reliable way to use the C flag is to have `_mm256_testc_si256` nearby, so that they share the instruction.

**Comment by topperc** [comment]
> > I think the tricky bit is that it isn't commutable if the C flag is used.
> 
> Yes, but I think the only reliable way to use the C flag is to have `_mm256_testc_si256` nearby, so that they share the instruction.

I meant it as a compiler implementation problem. We lose knowledge of whether _mm256_testc_si256 or _mm256_testz_si256 was used before we get to the place where commuting decisions happen. So we have to scan the users to check what flags they use.
